File: Animal/Chicken.cpp
/**
 * @file Chicken.cpp
 * @author 13515125/Muhammad Rafid Amrullah, 13517092/Michael Ray, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * @brief file berisi implementasi class Chicken
 * @version 0.1
 * @date 2019-03-20
 */

/* NIM/Nama  : 13515125/Muhammad Rafid Amrullah, 13517092/Michael Ray, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * Nama file : Chicken.cpp
 * Topik     : Tugas Besar 1 IF2210 - Pemograman Berorientasi Objek
 * Tanggal   : 20 Maret 2019
 * Deskripsi : Implementasi Chicken.cpp */

#include "Chicken.hpp"
// #include "ChickenEgg.hpp"
// #include "ChickenMeat.hpp"
#include <string.h>
#include <random>
#include <iostream>
using namespace std;

int Chicken::n_chicken = 0;

/**
 * @brief Construct a new Chicken object
 * 
 */
Chicken::Chicken(int id, int x, int y)
{
    this->id = id;
    this->voice = new char[15];
    strcpy(this->voice, "Petok");
    this->hungry = false;
    this->umur = 30;
    this->x = x;
    this->y = y;
    n_chicken++;
}

/**
 * @brief Destroy the Chicken object
 * 
 */
Chicken::~Chicken()
{
    delete [] voice;
    cout << "Chicken " << this->id << "is dead." << endl;
}

/**
 * @brief Get the Name object
 * 
 * @return char* 
 */
int Chicken::getId() const
{
    return this->id;
}

/**
 * @brief Get the Hungry object
 * 
 * @return boolean 
 */
bool Chicken::getHungry() const
{
    return this->hungry;
}

/**
 * @brief Set the Name object
 * 
 */
void Chicken::setId(int id)
{
    this->id = id;
}

/**
 * @brief Get the X object
 * 
 * @return int 
 */
int Chicken::getX()
{
    return this->x;
}

/**
 * @brief Get the Y object
 * 
 * @return int 
 */
int Chicken::getY()
{
    return this->y;
}

/**
 * @brief Function to get ChickenEgg
 * 
 */
void Chicken::interactProduct()
{
    // ChickenEgg();
}

/**
 * @brief Function to get ChickenMeat
 * 
 */
void Chicken::killProduct()
{
    // ChickenMeat();
}

/**
 * @brief Method for the animal to eat
 * 
 */
void Chicken::eat()
{
    this->hungry = false;
}

/**
 * @brief Method for the animal to move
 * 
 */
void Chicken::move(char** map, int row, int col)
{
    int turn;
    char next = '.';

    random_device dev;
    mt19937 rng(dev());
    uniform_int_distribution<mt19937::result_type> dist6(1,4);
    turn = dist6(rng);

    if (turn == 1)
    {
        if(this->y + 1 < row)
            next = map[this->y+1][this->x];
        if (next == 'o' or next == '*')
        {
            this->y++;
        }
    }
    else if (turn == 2)
    {
        if(this->x + 1 < col)
            next = map[this->y][this->x+1];
        if (next == 'o' or next == '*')
        {
            this->x++;
        }
    }
    else if (turn == 3)
    {
        if(this->y - 1 >= 0)
            next = map[this->y-1][this->x];
        if (next == 'o' or next == '*')
        {
            this->y--;
        }
    }
    else if (turn == 4)
    {
        if(this->x - 1 >= 0)
            next = map[this->y][this->x-1];
        if (next == 'o' or next == '*')
        {
            this->x--;
        }
    }
}

/**
 * @brief Method for the animal to voice
 * 
 */
void Chicken::sound()
{
    cout << this->voice << endl;
}

/**
 * @brief Method to render the animal to map
 * 
 * @return char 
 */
char Chicken::render()
{
    if (this->hungry)
    {
		return 'c';
	}
	else
	{
		return 'C';
	}
}

File: Java/model/animal/Chicken.java
/**
 * Chicken.java
 * <p>
 * File berisi class Chicken
 * </p>
 * @author 13515125/Muhammad Rafid Amrullah, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * @version 2.0
 * @since 2019-04-21
 */

package animal;

import java.io.*;
import java.lang.*;
import java.util.*;

/**
 * <p>
 * Class for rendering object to Chicken
 * </p>
 */
public class Chicken implements IProducing, KProducing
{
	private static int n_chicken = 0;
	private int id;
	private String voice;
	private boolean hunger, starvation, interactivity, produceMeat;
	private int x, y, tick;
	
	/**
	 * Chicken constructor
	 * @param y
	 * @param x
	 */
	public Chicken(int y, int x){
		this.id = ++n_chicken;
		this.voice = "Petok";
		this.hunger = false;
		this.y = y;
		this.x = x;
		this.tick = 0;
		this.starvation = false;
		this.interactivity = true;
		this.produceMeat = true;
	}
	
	/**
	 * Get the object id
	 * @return int
	 */
	public int getId(){
		return id;
	}

	/**
	 * Get the object hunger status
	 * @return boolean
	 */
	public boolean getHunger(){
		return hunger;
	}

	/**
	 * Get the object interactivity
	 * @return boolean
	 */
	public boolean getInteractivity(){
		return interactivity;
	}

	/**
	 * Get the object starve status
	 * @return boolean
	 */
	public boolean getStarvation(){
		return starvation;
	}

	/**
	 * Get the object kill status
	 * @return boolean
	 */
	public boolean isMeatProducing(){
		return produceMeat;
	}

	/**
	 * Get the object x position
	 * @return int
	 */
	public int getX(){
		return this.x;
	}

	/**
	 * Get the object y position
	 * @return int
	 */
	public int getY(){
		return this.y;
	}

	/**
	 * Eat function
	 */
	public void eat(){
		hunger = false;
		interactivity = true;
		tick = 0;
	}

	/**
	 * Move function
	 * @param map
	 */
	public void move(Character[][] map){
		int row = map.length;
		int col = map[0].length;
	    char next = '.';

	    Random random = new Random();
	    int turn = random.nextInt(5);

	    if (turn == 1)
	    {
	        if(this.y + 1 < row)
	            next = map[this.y+1][this.x];
	        if (next == 'o' || next == '*' || next == 'x' || next == '@')
	        {
	            this.y++;
	        }
	    }
	    else if (turn == 2)
	    {
	        if(this.x + 1 < col)
	            next = map[this.y][this.x+1];
	        if (next == 'o' || next == '*' || next == 'x' || next == '@')
	        {
	            this.x++;
	        }
	    }
	    else if (turn == 3)
	    {
	        if(this.y - 1 >= 0)
	            next = map[this.y-1][this.x];
	        if (next == 'o' || next == '*' || next == 'x' || next == '@')
	        {
	            this.y--;
	        }
	    }
	    else if (turn == 4)
	    {
	        if(this.x - 1 >= 0)
	            next = map[this.y][this.x-1];
	        if (next == 'o' || next == '*' || next == 'x' || next == '@')
	        {
	            this.x--;
	        }
	    }

	    boolean validTile = map[this.y][this.x] != 'o' && map[this.y][this.x] != '*';
	    validTile = validTile && map[this.y][this.x] != 'x' && map[this.y][this.x] != '@';

	    if(validTile)
	    	move(map);

	    if(tick == 5)
	        hunger = true;
	    else if(tick >= 25 && hunger)
	        starvation = true;

	    tick++;
	}

	/**
	 * Get the object sound
	 * @return String
	 */
	public String sound(){
		return new String("Chicken " + getId() + ": " + voice);
	}

	/**
	 * Get the object symbol
	 * @return char
	 */
	public char render(){
		return (hunger)? 'c' : 'C';
	}

	/**
	 * Interact function
	 */
	public void interact(){
		interactivity = false;
	}

	/**
	 * Kill function
	 * @return String
	 */
	public String kill(){
		return new String("Chicken " + getId() + " is killed");
	}
}
